### PR TITLE
Moving examples to use the new discrete spaces

### PR DESCRIPTION
**Pull request by quaquel**
This PR moves over all relevant examples to the experimental grid spaces. While making this move, the experimental grid spaces continue to be refined.

### CODE REVIEW

**Comment by EwoutH** [comment]
If you change this line:
https://github.com/projectmesa/mesa-examples/blob/2b4354481c6f0341ced94526f29035840d326adb/.github/workflows/test_examples.yml#L58

To:
```
pip install -U git+https://github.com/quaquel/mesa@move_spaces#egg=mesa
```
You can see if it works in CI.

**Comment by quaquel** [comment]
I am testing locally against my branch for #2286, but I'll update the the yml in this branch to show CI here as well.

**Comment by EwoutH** [comment]
This PR just need the imports switched right?

**Comment by quaquel** [comment]
The smart thing to do would be to change the imports to `experimental.cell_space`. This would make this pr compatible with main.

**Comment by EwoutH** [comment]
Awesome, sounds like a plan!

**Comment by EwoutH** [comment]
If you need anything on this PR let me know!

**Comment by quaquel** [comment]
> If you need anything on this PR let me know!

How to get the build to succeed? It now errors out on reaction?

Other than that, I think this is another half our of work later today and then its done.

**Comment by EwoutH** [comment]
Awesome progress!

> How to get the build to succeed? It now errors out on reaction?

I think the error is this init (https://github.com/projectmesa/mesa/blob/main/mesa/experimental/__init__.py) in combination with that we now don’t install Solara by default anymore.

File: examples/hotelling_law/hotelling_law/model.py
Comment (EwoutH) [inline examples/hotelling_law/hotelling_law/model.py]: Is there a space too much here?

**Review by EwoutH — COMMENTED** [review]
I really like how clean the API is looking, and how easy old components are replaced. Good stuff!

File: examples/hotelling_law/hotelling_law/model.py
Comment (quaquel) [reply examples/hotelling_law/hotelling_law/model.py]: I'll run ruff once I am done and fix it all

**Comment by quaquel** [comment]
> I really like how clean the API is looking, and how easy old components are replaced. Good stuff!

I am also pleasantly surprised by (1) how easy the switch is; and (2) how, in many places, it reduces code complexity.

**Comment by EwoutH** [comment]
Really awesome, I see some really neat changes in here.

Curious, did you encounter any limitations or go any ideas to do things more elegantly in the cell space?

Let's discuss Wednesday how we want to structure the examples for experimental features. Would be awesome if we can stabilize those together with features in future PRs, once the core examples are moved back.

**Comment by quaquel** [comment]
> Curious, did you encounter any limitations or go any ideas to do things more elegantly in the cell space?

I have made various PRs related to grid spaces while working on this. For example, the way we resolved movement [#2333](https://github.com/projectmesa/mesa/pull/2333) and [#2296](https://github.com/projectmesa/mesa/pull/2296), the split of neighborhood and get_neigborhood [#2309](https://github.com/projectmesa/mesa/pull/2309), and the width/height properties [#2348](https://github.com/projectmesa/mesa/pull/2348) all emerged out of making these examples work with experimental grid spaces. Some of these examples can be refined even further by using property layers (e.g., sugarscape) or using some of the features introduced in the aforementioned PRs. But I am quite happy with how expressive and complete the grid spaces are

**Comment by quaquel** [comment]
I have updated Schelling after #222.  @EwoutH any further comments or can this be approved?

**Comment by EwoutH** [comment]
Could you split of the changes in the "basic" (as in https://github.com/projectmesa/mesa/pull/2349#issuecomment-2408945055) examples into a new, separate branch and save that somewhere where you can find it back in a few weeks/months? Then we will merge those during/after cell space stabilization.

All "complex" and "showcase" examples can stay in this PR, then we will merge that ASAP.

Sorry it's a bit of a messy process, your effort is appreciated, and it looks like we now have a clear policy for experimental features in examples.

**Comment by quaquel** [comment]
> Could you split of the changes in the "basic" (as in https://github.com/projectmesa/mesa/pull/2349#issuecomment-2408945055) examples into a new, separate branch and save that somewhere where you can find it back in a few weeks/months? Then we will merge those during/after cell space stabilization.

It will be annoying but doable if I have time in a few days. Would it not be simpler to move the basis versions over to that PR? If I am not mistaken, only Schelling is missing at the moment. Next, we merge this, continue the move of advanced examples and once that PR is merged remove all moved over examples from mesa-examples.

**Comment by EwoutH** [comment]
For a proper port of the git history its way easier if we can move the examples at once, otherwise we're merging multiple related git histories which makes it even more complex.

I reverted the basic examples, and saved the changes in a backup branch: https://github.com/projectmesa/mesa-examples/tree/pr_198_examples_backup

https://github.com/projectmesa/mesa/pull/2357 should unblock the CI and then we're good to go.

**Comment by quaquel** [comment]
> For a proper port of the git history its way easier if we can move the examples at once, otherwise we're merging multiple related git histories which makes it even more complex.

Fair point. I had not considered that.

**Comment by EwoutH** [comment]
Main CI passes, awesome.

Thanks a lot for this huge effort!

**Review by EwoutH — APPROVED** [review]

File: examples/epstein_civil_violence/epstein_civil_violence/agent.py
Comment (EwoutH) [inline examples/epstein_civil_violence/epstein_civil_violence/agent.py:14]: Vision is something we might want to build in at some point. NetLogo has quite extensive support for it, for example.

File: examples/epstein_civil_violence/epstein_civil_violence/model.py
Comment (EwoutH) [inline examples/epstein_civil_violence/epstein_civil_violence/model.py:74]: Can we make properties for this? Like `cell.x`, `cell.y` and `cell.z`? Similar like https://github.com/projectmesa/mesa/pull/2348 but for the cell.

File: examples/hotelling_law/hotelling_law/agents.py
Comment (EwoutH) [inline examples/hotelling_law/hotelling_law/agents.py:165]: Is this correct? Why should is now be an AgentSet, when it first was a string (`model.consumer_preferences="default",`)

File: examples/hotelling_law/hotelling_law/model.py
Comment (EwoutH) [inline examples/hotelling_law/hotelling_law/model.py:193]: You can now do!
```suggestion
            x = self.random.randrange(self.grid.width)
            y = self.random.randrange(self.grid.height)
```

File: examples/hotelling_law/hotelling_law/model.py
Comment (EwoutH) [inline examples/hotelling_law/hotelling_law/model.py:204]: Don't we have an function to select a random cell?

File: examples/schelling_experimental/model.py
Comment (EwoutH) [inline examples/schelling_experimental/model.py:1]: Can be removed I think

File: examples/shape_example/shape_example/model.py
Comment (EwoutH) [inline examples/shape_example/shape_example/model.py:25]:
```suggestion
            x = self.random.randrange(self.grid.width)
            y = self.random.randrange(self.grid.height)
```

**Review by EwoutH — COMMENTED** [review]
I was a bit enthosiastic on my approval, only after realizing I didn't do a proper code review yet.

Overal, it looks amazing, with some really nice cleanups and elegant constructs! I have a few small things, then can (but should) be dealt with in a follow-up PR.

Especially check if if all your changes were ported correctly after https://github.com/projectmesa/mesa-examples/pull/220, and then remove the duplicate files.

File: examples/hotelling_law/hotelling_law/agents.py
Comment (quaquel) [reply examples/hotelling_law/hotelling_law/agents.py:165]: this is indeed probably a mistake on my end

**Comment by quaquel** [comment]
Thanks for these comments. I just like to point out that most of these are comments on the original code from which I started, not the changes made in this PR. Some of the other comments are about using features that were added based on what was learned while doing this. I'll try to go through these in follow up PRs when I have time. A quick look suggests only 1 is pointing out a mistake in my code.

**Comment by EwoutH** [comment]
Oh certainly, and none of them were blocking. Just things I noticed when going through the diff.

Do you prefer to fix some of them before the example move, or after?

File: examples/schelling_experimental/model.py
Comment (quaquel) [reply examples/schelling_experimental/model.py:1]: why?

File: examples/bank_reserves/BankReservesModel_Data.csv
Comment (quaquel) [reply examples/bank_reserves/BankReservesModel_Data.csv:1]: not sure, but the entire batch run example stuff seems incorrect so this needs a bit more attention

File: examples/aco_tsp/aco_tsp/model.py
Comment (quaquel) [reply examples/aco_tsp/aco_tsp/model.py:97]: not sure what you are proposing. This is in the agent class.

**Pull request merged**
